voice/src/main/kotlin/udp/DefaultVoiceUdpConnection.kt

```diff
@@ -0,0 +1,66 @@
+package dev.kord.voice.udp
+
+import io.ktor.network.selector.*
+import io.ktor.network.sockets.*
+import io.ktor.util.network.*
+import io.ktor.utils.io.core.*
+import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.*
+import mu.KotlinLogging
+import kotlin.coroutines.CoroutineContext
+
+private val udpConnectionLogger = KotlinLogging.logger { }
+
+data class DefaultVoiceUdpConnectionData(
+    val dispatcher: CoroutineDispatcher
+)
+
+internal class DefaultVoiceUdpConnection(
+    val data: DefaultVoiceUdpConnectionData
+) : VoiceUdpConnection, CoroutineScope {
+    override val coroutineContext: CoroutineContext =
+        SupervisorJob() + data.dispatcher + CoroutineName("Discord Voice UDP Connection")
+
+    private lateinit var socket: ConnectedDatagramSocket
+    private lateinit var configuration: VoiceUdpConnectionConfiguration
```

Comment:
This is not thread safe, there is no happens-before relationship between setting and using `socket` and `configuration` (between `start`, `discoverIp` and `send`).
-> `@Volatile` or `atomic`?

Reply:
The lateinit socket is something done in Gateway and VoiceGateway, I think it would be best to rework the concurrency structure of each gateway/connection in its own pr, which *I think* is being done.

i don't mind making configuration atomic, however.